src/duckdb.ts

```diff
@@ -0,0 +1,97 @@
+import {existsSync} from "node:fs";
+import {mkdir, writeFile} from "node:fs/promises";
+import {dirname, join} from "node:path/posix";
+import {cross} from "d3-array";
+import type {DuckDBConfig} from "./config.js";
+import {faint} from "./tty.js";
+
+const downloadRequests = new Map<string, Promise<string>>();
+
+export async function duckDBManifest(
+  duckdb: DuckDBConfig,
+  {root, log, aliases}: {root: string; log?: boolean; aliases?: Map<string, string>}
+) {
+  return {
+    bundles: duckdb.bundles,
+    extensions: await Promise.all(
+      cross(duckdb.bundles, duckdb.install).map(async ([p, name]) => {
+        let ext = await resolveDuckDBExtension(root, p, duckdb.source[name], name);
+        if (aliases?.has(ext)) ext = aliases.get(ext)!;
+        return [
+          name,
+          {
+            ref: dirname(dirname(dirname(ext))),
+            load: duckdb.load.includes(name),
+            bundle: p
+          }
+        ];
+      })
+    ),
+    log
+  };
+}
+
+/**
+ * Given a duckdb configuration and an extension name such as "parquet", saves
+ * the binary to _duckdb/{hash}/v1.1.1/wasm_{p}/parquet.duckdb_extension.wasm
+ * for every supported platform p ("eh" and "mvp"), and returns a content-hashed
+ * reference (_duckdb/{hash}) to use in the corresponding DuckDB INSTALL
+ * statement. The repo is structured as required by DuckDB with:
+ * ${repo}/v1.1.1/wasm_{p}/${name}.duckdb_extension.wasm
+ */
+export async function resolveDuckDBExtension(root: string, p: string, repo: string, name: string): Promise<string> {
+  if (!repo.startsWith("https://")) throw new Error(`invalid repo: ${repo}`);
+  const cache = join(root, ".observablehq", "cache");
+  const file = `${name}.duckdb_extension.wasm`;
+  const ref = `${repo}/v1.1.1/wasm_${p}/${file}`.slice("https://".length);
```

Comment:
Where does the `v1.1.1` come from? Will this need to change when new releases of DuckDB come out? I would like to minimize the number of things we need to maintain and ideally allow Framework to pick up new releases of DuckDB automatically. If we bake-in versions here then we become the bottleneck for upgrading. The main reason we pinned the version of DuckDB-Wasm before was they made the mistake of publishing the prerelease of 1.29.0 under the `latest` disttag (and didn’t correct it for months). It’s great that we’re adding self-hosting of DuckDB extensions to Framework, but otherwise I’d still like to minimize how much we’re “in the loop” for new DuckDB-Wasm releases.

Reply:
Given the history here, I think we're better off manually locking the versions. (Agree that, ideally, we wouldn't have to.) I think `v1.1.1` is the underlying DuckDB version, but I can't find the explanation in the docs.

Reply:
Not sure if there is any way other than `SELECT VERSION()` to get the version of DuckDB that DuckDB-Wasm uses under the hood… and we’d need to make sure that we’re using the right version of DuckDB-Wasm when doing the resolution here. So maybe pinning both is the only viable option for now. Still, we could at least pull the `v1.1.1` up to a global variable, and keep the DuckDB and DuckDB-Wasm versions defined in the same place so that we’re more likely to update both of them at the same time.

Reply:
> it’s the DuckDB version as described [in the DuckDB docs](https://duckdb.org/docs/extensions/overview.html#installation-location)

I think it might be slightly different for duckdb-wasm. For stable versions it's certainly going to follow the same pattern, but for nightly builds I'm not sure how the URL is built.

For now, when I upgrade to a nightly by changing `1.29.0` to `1.29.1-dev3.0` (which is built against DuckDB `1.1.2`, I can see that the client is trying to load
`http://127.0.0.1:3000/_duckdb/extensions.duckdb.org/v1.1.2/wasm_eh/parquet.duckdb_extension.wasm` (which is not available in framework since `1.1.1` is hard-coded).

I've now centralized the version numbers to make this easier to update in the future.

Reply:
I've tried to make the duckdb-wasm and DuckDB versions part of the configuration (to allow users to experiment with upgrades), but this seems to require deep changes, that I'm not ready to do. I've left this out for now (and note that it might be solved by an environment variable instead).